pkg/secrethub/auth_aws.go

```diff
@@ -0,0 +1,62 @@
+package secrethub
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/secrethub/secrethub-go/internals/api"
+	"github.com/secrethub/secrethub-go/internals/auth"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+)
+
+type awsAuthService struct {
+	client client
+}
+
+func newAWSAuthService(client client) AuthMethodService {
+	return &awsAuthService{
+		client: client,
+	}
+}
+
+func (s awsAuthService) Authenticate() (auth.Authenticator, error) {
```

Comment:
Add godoc comments por favor

Reply:
I feel we need to look at the contract that this exposes. In my mind, I should be able to say something like:

```go
client := secrethub.NewClient()
client := secrethub.NewClient().WithAWS(foo)?
client := secrethub.NewClientWithAWS(foo)?
```

Reply:
Consider making it private or creating a distinction between plain clients and authenticated clients, e.g. 

```go
func (c client) CreateSession(foo) Client {
...
}
```

Reply:
Absolutely agree that the contract can be improved. If we want `client := secrethub.NewClient().WithAWS(foo)` then how do we get a client with file-based credentials? `client := secrethub.NewClient().WithFileCredential("path")`?

Reply:
Jotted down some notes as to how we could use functional options like we discussed in person. Not really sure whether I like the `client.With(opts)` function but it has a certain elegance. It's pretty similar to the pattern a lot of Golang router packages use, e.g. `chi.Router.Use`.


```go
package secrethub

type Option func(c Client) (Client, error)

func NewClient(opts ...Option) (Client, error) {
   ...
   c := foo 
   for _, opt := range opts {
      c, err = opt(c)
      if err != nil {
         return nil, err
      }
   }

   return c, nil
}

func (c client) With(opts ...Option) (Client, error) { ... }

func WithServerURL(url url.URL) Option {...}

func WithTransport(t http.Transport) Option {...}

func WithAWS(foo string) Option {
   return func(c Client) (Client, error) {
      // do aws stuff
      ...
   }
} 

func WithCredential(raw string, passphrase string) Option {
   return func(c Client) (Client, error) {
      // parse credential etc
      ...
   }
}

func WithCredentialFile(path string) Option {}
```

So you could call it something like:

```go
// Create a new client
client, err := secrethub.NewClient(
   secrethub.WithAWS(foo), 
   secrethub.WithServerURL("https://api.secrethub.io"),
)

// Change it on the fly, e.g. for a handful of requests
specialClient, err = client.With(secrethub.WithTransport(t))
```